File: ios/Sources/KlinikAuthFeature/AccountScreen.swift
import SwiftUI
import KlinikAPI
import KlinikCore
import KlinikDesign

/// Where somebody sees which devices are signed in, and ends the ones that
/// should not be (spec M1).
public struct AccountScreen: View {
    @Environment(\.colorScheme) private var scheme

    private let model: AccountModel
    private let signOut: () async -> Void
    /// The biometric setting, owned by the shell. Nil on a platform that has
    /// no such thing, which leaves the section out rather than showing a
    /// switch that does nothing.
    private let biometrics: BiometricSetting?

    /**
     * Whether this account may turn its second factor off.
     *
     * False for clinic staff, for whom it is mandatory — the server refuses,
     * and a button that always fails is worse than no button. Passed in
     * because the shell knows the role and this screen does not.
     */
    private let canDisableTwoFactor: Bool

    /**
     * Producing the patient's own copy of their record (KVKK m.11).
     *
     * A closure returning the file, because writing it and presenting a share
     * sheet is the shell's business and fetching it is the API's. Nil for
     * staff: `me/data-export` is a patient's own record and there is none.
     */
    private let exportMyData: (() async -> URL?)?

    @State private var state = AccountState()
    @State private var confirmingSignOutEverywhere = false
    @State private var changingPassword = false
    @State private var disablingTwoFactor = false
    @State private var twoFactorOff = false
    @State private var exporting = false
    @State private var exported: URL?
    @State private var exportFailed = false

    public init(
        model: AccountModel,
        signOut: @escaping () async -> Void,
        biometrics: BiometricSetting? = nil,
        canDisableTwoFactor: Bool = false,
        exportMyData: (() async -> URL?)? = nil
    ) {
        self.model = model
        self.signOut = signOut
        self.biometrics = biometrics
        self.canDisableTwoFactor = canDisableTwoFactor
        self.exportMyData = exportMyData
    }

    public var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: Tokens.Spacing.lg) {
                ErrorBanner(message: state.error)

                switch state.phase {
                case .loading:
                    VStack(spacing: Tokens.Spacing.lg) {
                        SkeletonCard(lines: 3)
                        SkeletonCard(lines: 3)
                    }
                    .accessibilityElement()
                    .accessibilityLabel(L10n.string("common.loading"))

                case .failed(let message):
                    MessageState(
                        icon: Tokens.State.labCritical.iconName,
                        text: message,
                        retryTitle: L10n.string("common.retry")
                    ) {
                        await reload()
                    }

                case .loaded:
                    security
                    credentials
                    myData
                    sessions
                }
            }
            .padding(Tokens.Spacing.lg)
        }
        .background(Tokens.Palette.background.resolve(for: scheme))
        .navigationTitle(L10n.string("account.title"))
        .refreshable { await reload() }
        .task { await reload() }
        .confirmationDialog(
            L10n.string("account.signOutEverywhereConfirm"),
            isPresented: $confirmingSignOutEverywhere,
            titleVisibility: .visible
        ) {
            Button(L10n.string("account.signOutEverywhere"), role: .destructive) {
                Task {
                    if await model.signOutEverywhere() { await signOut() }
                    state = model.currentState()
                }
            }

            Button(L10n.string("common.cancel"), role: .cancel) {}
        }
        .sheet(isPresented: $changingPassword) {
            ChangePasswordSheet { current, next in
                let changed = await model.changePassword(current: current, new: next)
                state = model.currentState()

                return changed
            } finished: {
                // Every session is dead, this one included. Staying would be a
                // screen where nothing works and nothing says why.
                changingPassword = false
                Task { await signOut() }
            }
        }
        .sheet(isPresented: $disablingTwoFactor) {
            DisableTwoFactorSheet { code in
                let disabled = await model.disableTwoFactor(code: code)
                state = model.currentState()

                if disabled {
                    twoFactorOff = true
                    disablingTwoFactor = false
                }

                return disabled
            }
        }
    }

    /**
     * The patient's own copy of their record (KVKK m.11).
     *
     * A right the privacy notice promises in writing, which until now had
     * nothing behind it in the app. Produced on demand and handed to the share
     * sheet rather than saved somewhere: it is the reader's file, and where it
     * goes is their decision.
     */
    @ViewBuilder
    private var myData: some View {
        if let exportMyData {
            VStack(alignment: .leading, spacing: Tokens.Spacing.md) {
                SectionHeader(title: L10n.string("dataExport.title"))

                Card {
                    VStack(alignment: .leading, spacing: Tokens.Spacing.sm) {
                        Text(L10n.string("dataExport.explain"))
                            .font(Tokens.Typography.calloutRelative)
                            .foregroundStyle(Tokens.Palette.textSecondary.resolve(for: scheme))
                            .fixedSize(horizontal: false, vertical: true)

                        if let exported {
                            ShareLink(item: exported) {
                                Label(
                                    L10n.string("dataExport.share"),
                                    systemImage: "square.and.arrow.up"
                                )
                                .font(Tokens.Typography.bodyRelative)
                                .frame(maxWidth: .infinity, minHeight: Tokens.minimumTouchTarget)
                            }
                            .foregroundStyle(Tokens.Palette.accent.resolve(for: scheme))
                        } else {
                            PrimaryButton(
                                title: L10n.string("dataExport.prepare"),
                                isBusy: exporting,
                                isEnabled: !exporting
                            ) {
                                exporting = true
                                exportFailed = false
                                exported = await exportMyData()
                                exportFailed = exported == nil
                                exporting = false
                            }
                        }

                        if exportFailed {
                            Text(L10n.string("dataExport.failed"))
                                .font(Tokens.Typography.captionRelative)
                                .foregroundStyle(Tone.critical.foreground.resolve(for: scheme))
                                .fixedSize(horizontal: false, vertical: true)
                        }
                    }
                }
            }
        }
    }

    /// Password and second factor — the two things about the account itself
    /// rather than about the devices it is signed in on.
    @ViewBuilder
    private var credentials: some View {
        VStack(alignment: .leading, spacing: Tokens.Spacing.md) {
            SectionHeader(title: L10n.string("account.security"))

            Card {
                VStack(alignment: .leading, spacing: Tokens.Spacing.sm) {
                    Button(L10n.string("account.changePassword")) { changingPassword = true }
                        .font(Tokens.Typography.bodyRelative)
                        .frame(maxWidth: .infinity, minHeight: Tokens.minimumTouchTarget,
                               alignment: .leading)
                        .foregroundStyle(Tokens.Palette.accent.resolve(for: scheme))

                    Text(L10n.string("account.changePasswordHint"))
                        .font(Tokens.Typography.captionRelative)
                        .foregroundStyle(Tokens.Palette.textSecondary.resolve(for: scheme))
                        .fixedSize(horizontal: false, vertical: true)
                }
            }

            twoFactor
        }
    }

    @ViewBuilder
    private var twoFactor: some View {
        Card {
            VStack(alignment: .leading, spacing: Tokens.Spacing.sm) {
                HStack {
                    Text(L10n.string("account.twoFactor"))
                        .font(Tokens.Typography.bodyRelative)
                        .foregroundStyle(Tokens.Palette.textPrimary.resolve(for: scheme))

                    Spacer(minLength: Tokens.Spacing.sm)

                    Badge(
                        L10n.string(twoFactorOff ? "common.off" : "account.twoFactorOn"),
                        tone: twoFactorOff ? .neutral : .success
                    )
                }

                if !canDisableTwoFactor {
                    // Said rather than hidden: a clinician who goes looking for
                    // this switch should find the reason, not an absence.
                    Text(L10n.string("account.twoFactorMandatory"))
                        .font(Tokens.Typography.captionRelative)
                        .foregroundStyle(Tokens.Palette.textSecondary.resolve(for: scheme))
                        .fixedSize(horizontal: false, vertical: true)
                } else if twoFactorOff {
                    Text(L10n.string("account.twoFactorDisabled"))
                        .font(Tokens.Typography.captionRelative)
                        .foregroundStyle(Tokens.Palette.textSecondary.resolve(for: scheme))
                        .fixedSize(horizontal: false, vertical: true)
                } else {
                    Button(L10n.string("account.disableTwoFactor")) { disablingTwoFactor = true }
                        .font(Tokens.Typography.bodyRelative)
                        .frame(maxWidth: .infinity, minHeight: Tokens.minimumTouchTarget,
                               alignment: .leading)
                        .foregroundStyle(Tone.critical.foreground.resolve(for: scheme))
                }
            }
        }
    }

    @ViewBuilder
    private var security: some View {
        if let biometrics {
            VStack(alignment: .leading, spacing: Tokens.Spacing.md) {
                SectionHeader(title: L10n.string("account.security"))

                Card {
                    VStack(alignment: .leading, spacing: Tokens.Spacing.sm) {
                        if biometrics.isAvailable {
                            Toggle(
                                L10n.string("biometrics.enable"),
                                isOn: Binding(
                                    get: { biometrics.isEnabled() },
                                    set: { biometrics.setEnabled($0) }
                                )
                            )
                            .font(Tokens.Typography.bodyRelative)
                            .frame(minHeight: Tokens.minimumTouchTarget)

                            Text(L10n.string("biometrics.enableHint"))
                                .font(Tokens.Typography.captionRelative)
                                .foregroundStyle(
                                    Tokens.Palette.textSecondary.resolve(for: scheme)
                                )
                                .fixedSize(horizontal: false, vertical: true)
                        } else {
                            Text(L10n.string("biometrics.unavailable"))
                                .font(Tokens.Typography.bodyRelative)
                                .foregroundStyle(
                                    Tokens.Palette.textSecondary.resolve(for: scheme)
                                )
                                .fixedSize(horizontal: false, vertical: true)
                        }
                    }
                }
            }
        }
    }

    @ViewBuilder
    private var sessions: some View {
        if let current = state.current {
            VStack(alignment: .leading, spacing: Tokens.Spacing.md) {
                SectionHeader(title: L10n.string("account.thisDevice"))
                card(current, canRevoke: false)
            }
        }

        VStack(alignment: .leading, spacing: Tokens.Spacing.md) {
            SectionHeader(
                title: L10n.string("account.otherDevices"),
                subtitle: "\(state.others.count)"
            )

            if state.others.isEmpty {
                Card {
                    Text(L10n.string("account.noOtherDevices"))
                        .font(Tokens.Typography.bodyRelative)
                        .foregroundStyle(Tokens.Palette.textSecondary.resolve(for: scheme))
                }
            } else {
                ForEach(state.others) { session in
                    card(session, canRevoke: true)
                }
            }
        }

        Button(L10n.string("account.signOutEverywhere")) {
            confirmingSignOutEverywhere = true
        }
        .font(Tokens.Typography.calloutRelative)
        .frame(maxWidth: .infinity, minHeight: Tokens.minimumTouchTarget)
        .foregroundStyle(Tokens.Palette.critical.resolve(for: scheme))
    }

    private func card(_ session: DeviceSession, canRevoke: Bool) -> some View {
        Card(tone: session.current ? .info : .neutral) {
            VStack(alignment: .leading, spacing: Tokens.Spacing.md) {
                HStack(spacing: Tokens.Spacing.md) {
                    Image(systemName: AccountScreen.symbol(for: session.platform))
                        .font(Tokens.Typography.headingRelative)
                        .frame(width: 32)
                        .foregroundStyle(Tokens.Palette.accent.resolve(for: scheme))
                        .accessibilityHidden(true)

                    VStack(alignment: .leading, spacing: Tokens.Spacing.xxs) {
                        Text(session.deviceName ?? L10n.string("account.unknownDevice"))
                            .font(Tokens.Typography.subheadingRelative)
                            .foregroundStyle(Tokens.Palette.textPrimary.resolve(for: scheme))

                        Text(
                            String(
                                format: L10n.string("account.lastSeen"),
                                session.lastSeenAt.formatted(date: .abbreviated, time: .shortened)
                            )
                        )
                        .font(Tokens.Typography.captionRelative)
                        .foregroundStyle(Tokens.Palette.textSecondary.resolve(for: scheme))
                    }

                    Spacer(minLength: 0)

                    if session.current {
                        Badge(
                            L10n.string("account.thisDevice"),
                            tone: .info,
                            symbol: "checkmark.circle"
                        )
                    }
                }
                .accessibilityElement(children: .combine)

                // Shown because it is how somebody recognises a session they do
                // not remember. It is the clinic's own record, not a lookup.
                if let ip = session.ipAddress {
                    FieldRow(label: L10n.string("account.address"), value: ip)
                }

                if canRevoke {
                    Button(L10n.string("account.revoke")) {
                        Task {
                            await model.revoke(session.familyId)
                            state = model.currentState()
                        }
                    }
                    .font(Tokens.Typography.calloutRelative)
                    .frame(maxWidth: .infinity, minHeight: Tokens.minimumTouchTarget)
                    .foregroundStyle(Tokens.Palette.critical.resolve(for: scheme))
                    .disabled(state.busyId != nil)
                }
            }
        }
    }

    static func symbol(for platform: String?) -> String {
        switch platform?.lowercased() {
        case "ios": return "iphone"
        case "android": return "candybarphone"
        case "web": return "desktopcomputer"
        default: return "questionmark.circle"
        }
    }

    private func reload() async {
        await model.load()
        state = model.currentState()
    }
}


/**
 * The biometric setting, as this screen needs it.
 *
 * A small struct of closures rather than the lock itself: the account screen
 * lives in the auth feature, the lock lives in the shell, and a feature module
 * that imported the shell would make the dependency graph a circle.
 */
public struct BiometricSetting: Sendable {
    public let isAvailable: Bool
    public let isEnabled: @MainActor @Sendable () -> Bool
    public let setEnabled: @MainActor @Sendable (Bool) -> Void

    public init(
        isAvailable: Bool,
        isEnabled: @escaping @MainActor @Sendable () -> Bool,
        setEnabled: @escaping @MainActor @Sendable (Bool) -> Void
    ) {
        self.isAvailable = isAvailable
        self.isEnabled = isEnabled
        self.setEnabled = setEnabled
    }
}

/**
 * Changing the password.
 *
 * Confirmed twice, because the field is masked and a typo here locks somebody
 * out of their own clinical record until the clinic resets it by hand. The
 * rules are shown as they are broken rather than after the server refuses.
 */
struct ChangePasswordSheet: View {
    @Environment(\.colorScheme) private var scheme
    @Environment(\.dismiss) private var dismiss

    /// Returns true when the server accepted it.
    let change: (String, String) async -> Bool
    /// Called after the reader has read what happens next.
    let finished: () -> Void

    @State private var current = ""
    @State private var next = ""
    @State private var confirmation = ""
    @State private var busy = false
    @State private var done = false

    private var problems: [String] {
        next.isEmpty ? [] : PasswordRules.problems(with: next)
    }

    private var mismatched: Bool {
        !confirmation.isEmpty && confirmation != next
    }

    private var canSubmit: Bool {
        !current.isEmpty && !next.isEmpty && confirmation == next && problems.isEmpty && !busy
    }

    var body: some View {
        NavigationStack {
            ScrollView {
                VStack(alignment: .leading, spacing: Tokens.Spacing.lg) {
                    if done {
                        Card(tone: .success) {
                            Text(L10n.string("account.passwordChanged"))
                                .font(Tokens.Typography.bodyRelative)
                                .fixedSize(horizontal: false, vertical: true)
                        }

                        PrimaryButton(
                            title: L10n.string("common.close"),
                            isBusy: false,
                            isEnabled: true
                        ) {
                            finished()
                        }
                    } else {
                        form
                    }
                }
                .padding(Tokens.Spacing.lg)
            }
            .background(Tokens.Palette.background.resolve(for: scheme))
            .navigationTitle(L10n.string("account.changePassword"))
            .toolbar {
                if !done {
                    ToolbarItem(placement: .cancellationAction) {
                        Button(L10n.string("common.cancel")) { dismiss() }
                    }
                }
            }
        }
    }

    @ViewBuilder
    private var form: some View {
        Text(L10n.string("account.changePasswordHint"))
            .font(Tokens.Typography.calloutRelative)
            .foregroundStyle(Tokens.Palette.textSecondary.resolve(for: scheme))
            .fixedSize(horizontal: false, vertical: true)

        LabelledField(
            label: L10n.string("account.currentPassword"),
            text: $current,
            isSecure: true,
            contentType: .password,
            keyboard: .default
        )

        LabelledField(
            label: L10n.string("account.newPassword"),
            text: $next,
            isSecure: true,
            contentType: .newPassword,
            keyboard: .default
        )

        LabelledField(
            label: L10n.string("auth.confirmPassword"),
            text: $confirmation,
            isSecure: true,
            contentType: .newPassword,
            keyboard: .default
        )

        if mismatched {
            Text(L10n.string("auth.passwordsDiffer"))
                .font(Tokens.Typography.captionRelative)
                .foregroundStyle(Tone.critical.foreground.resolve(for: scheme))
        }

        PasswordRuleList(problems: problems, showAll: next.isEmpty)

        PrimaryButton(
            title: L10n.string("account.changePassword"),
            isBusy: busy,
            isEnabled: canSubmit
        ) {
            busy = true
            done = await change(current, next)
            busy = false
        }
    }
}

/**
 * Turning the second factor off.
 *
 * A current code is required, so somebody who has the phone in their hand can
 * do it and somebody who merely has the password cannot.
 */
struct DisableTwoFactorSheet: View {
    @Environment(\.colorScheme) private var scheme
    @Environment(\.dismiss) private var dismiss

    let disable: (String) async -> Bool

    @State private var code = ""
    @State private var busy = false

    var body: some View {
        NavigationStack {
            ScrollView {
                VStack(alignment: .leading, spacing: Tokens.Spacing.lg) {
                    Text(L10n.string("account.disableTwoFactorHint"))
                        .font(Tokens.Typography.calloutRelative)
                        .foregroundStyle(Tokens.Palette.textSecondary.resolve(for: scheme))
                        .fixedSize(horizontal: false, vertical: true)

                    LabelledField(
                        label: L10n.string("account.twoFactorCode"),
                        text: $code,
                        isSecure: false,
                        contentType: .oneTimeCode,
                        keyboard: .numberPad
                    )

                    PrimaryButton(
                        title: L10n.string("account.disableTwoFactor"),
                        isBusy: busy,
                        isEnabled: code.count == DisableTwoFactorSheet.codeLength && !busy
                    ) {
                        busy = true
                        _ = await disable(code)
                        busy = false
                    }
                }
                .padding(Tokens.Spacing.lg)
            }
            .background(Tokens.Palette.background.resolve(for: scheme))
            .navigationTitle(L10n.string("account.disableTwoFactor"))
            .toolbar {
                ToolbarItem(placement: .cancellationAction) {
                    Button(L10n.string("common.cancel")) { dismiss() }
                }
            }
        }
    }

    /// Six digits, as every TOTP code is. `nonisolated` because a static on a
    /// `View` otherwise inherits the view's main-actor isolation.
    nonisolated static let codeLength = 6
}

File: android/feature/account-ui/src/main/kotlin/xyz/klinik/feature/account/ui/AccountScreen.kt
package xyz.klinik.feature.account.ui

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ColumnScope
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Button
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.heading
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.input.PasswordVisualTransformation
import xyz.klinik.design.Tokens
import xyz.klinik.design.klinikColor
import xyz.klinik.feature.account.AccountOutcome
import xyz.klinik.feature.account.AccountPhase
import xyz.klinik.feature.account.AccountState
import xyz.klinik.network.SessionSummary
import xyz.klinik.network.UiText
import xyz.klinik.shell.PasswordRules

/** Text the screen needs, resolved by the caller from string resources. */
data class AccountStrings(
    val title: String,
    val retry: String,
    val security: String,
    val changePassword: String,
    val changePasswordHint: String,
    val currentPassword: String,
    val newPassword: String,
    val passwordChanged: String,
    val twoFactor: String,
    val twoFactorOn: String,
    val twoFactorCode: String,
    val disableTwoFactor: String,
    val disableTwoFactorHint: String,
    val twoFactorDisabled: String,
    val twoFactorMandatory: String,
    val otherDevices: String,
    val noOtherDevices: String,
    val thisDevice: String,
    val unknownDevice: String,
    val lastSeen: (String) -> String,
    val revoke: String,
    val signOutEverywhere: String,
    val signOutEverywhereConfirm: String,
    val dataExport: String,
    val dataExportHint: String,
    val exportRows: (Int) -> String,
    val exportOmitted: String,
    val save: String,
    /** A password rule, with its number already in the sentence. */
    val rule: (PasswordRules.Problem) -> String,
    val message: (UiText) -> String,
)

/**
 * The account somebody signed in with (spec T7.3).
 *
 * Every action on this screen is irreversible in the same way, so each one
 * says what it will do before it does it: changing a password ends every
 * session including this one, and the sentence is above the fields rather than
 * in the confirmation afterwards. The password rules are shown while somebody
 * types, because a rejection that arrives from the network on a form filled in
 * twice reads as the app being broken.
 */
@Composable
fun AccountScreen(
    state: AccountState,
    strings: AccountStrings,
    onChangePassword: (current: String, next: String) -> Unit,
    onDisableTwoFactor: (String) -> Unit,
    onEndSession: (SessionSummary) -> Unit,
    onSignOutEverywhere: () -> Unit,
    onExport: () -> Unit,
    onSaveExport: (String) -> Unit,
    onRetry: () -> Unit,
    modifier: Modifier = Modifier,
) {
    Surface(color = klinikColor("background"), modifier = modifier.fillMaxSize()) {
        when (val phase = state.phase) {
            AccountPhase.Loading -> Centered { CircularProgressIndicator() }

            is AccountPhase.Failed -> Centered {
                Column(horizontalAlignment = Alignment.CenterHorizontally) {
                    Text(strings.message(phase.message), color = klinikColor("critical"))
                    TextButton(
                        onClick = onRetry,
                        modifier = Modifier.heightIn(min = Tokens.minimumTouchTarget),
                    ) {
                        Text(strings.retry)
                    }
                }
            }

            AccountPhase.Loaded -> Column(
                modifier = Modifier
                    .fillMaxSize()
                    .verticalScroll(rememberScrollState())
                    .padding(Tokens.Spacing.lg),
                verticalArrangement = Arrangement.spacedBy(Tokens.Spacing.lg),
            ) {
                Text(
                    strings.title,
                    fontSize = Tokens.Typography.heading.size,
                    fontWeight = Tokens.Typography.heading.weight,
                    color = klinikColor("textPrimary"),
                    modifier = Modifier.semantics { heading() },
                )

                state.identity?.let { identity ->
                    Text(identity.displayName, color = klinikColor("textSecondary"))
                }

                state.error?.let {
                    Text(strings.message(it), color = klinikColor("critical"))
                }

                Outcome(state, strings, onSaveExport)

                PasswordSection(state, strings, onChangePassword)
                TwoFactorSection(state, strings, onDisableTwoFactor)
                SessionsSection(state, strings, onEndSession, onSignOutEverywhere)

                Section(strings.dataExport) {
                    Text(strings.dataExportHint, color = klinikColor("textSecondary"))

                    OutlinedButton(
                        onClick = onExport,
                        enabled = !state.busy,
                        modifier = Modifier
                            .fillMaxWidth()
                            .heightIn(min = Tokens.minimumTouchTarget),
                    ) {
                        Text(strings.dataExport)
                    }
                }
            }
        }
    }
}

@Composable
private fun Outcome(
    state: AccountState,
    strings: AccountStrings,
    onSaveExport: (String) -> Unit,
) {
    when (val outcome = state.outcome) {
        null -> Unit

        // Said as an outcome rather than a toast: the session is gone and the
        // next thing this person does is sign in again, which needs
        // explaining.
        AccountOutcome.PasswordChanged -> Text(
            strings.passwordChanged,
            color = klinikColor("success"),
        )

        AccountOutcome.TwoFactorDisabled -> Text(
            strings.twoFactorDisabled,
            color = klinikColor("success"),
        )

        is AccountOutcome.Exported -> Column(
            verticalArrangement = Arrangement.spacedBy(Tokens.Spacing.sm),
        ) {
            Text(strings.exportRows(outcome.export.rowCount), color = klinikColor("textPrimary"))

            // What the file leaves out, named. A portability file with silent
            // gaps is worse than one that says where they are.
            if (outcome.export.notIncluded.isNotEmpty()) {
                Text(
                    strings.exportOmitted,
                    fontSize = Tokens.Typography.caption.size,
                    color = klinikColor("textSecondary"),
                )

                outcome.export.notIncluded.forEach { line ->
                    Text(
                        line,
                        fontSize = Tokens.Typography.caption.size,
                        color = klinikColor("warning"),
                    )
                }
            }

            Button(
                onClick = { onSaveExport(outcome.json) },
                modifier = Modifier
                    .fillMaxWidth()
                    .heightIn(min = Tokens.minimumTouchTarget),
            ) {
                Text(strings.save)
            }
        }
    }
}

@Composable
private fun PasswordSection(
    state: AccountState,
    strings: AccountStrings,
    onChangePassword: (String, String) -> Unit,
) {
    var current by remember { mutableStateOf("") }
    var next by remember { mutableStateOf("") }

    val problems = if (next.isEmpty()) emptyList() else PasswordRules.problems(next)

    Section(strings.changePassword) {
        // Before the fields, not after the button: signing every device out is
        // the kind of thing somebody should know is coming.
        Text(strings.changePasswordHint, color = klinikColor("textSecondary"))

        OutlinedTextField(
            value = current,
            onValueChange = { current = it },
            label = { Text(strings.currentPassword) },
            visualTransformation = PasswordVisualTransformation(),
            modifier = Modifier.fillMaxWidth(),
        )

        OutlinedTextField(
            value = next,
            onValueChange = { next = it },
            label = { Text(strings.newPassword) },
            visualTransformation = PasswordVisualTransformation(),
            modifier = Modifier.fillMaxWidth(),
        )

        // While typing, so nobody fills the form in twice to be told by the
        // network.
        problems.forEach { problem ->
            Text(
                strings.rule(problem),
                fontSize = Tokens.Typography.caption.size,
                color = klinikColor("warning"),
            )
        }

        Button(
            onClick = {
                onChangePassword(current, next)
                current = ""
                next = ""
            },
            enabled = !state.busy && current.isNotBlank() && problems.isEmpty() && next.isNotEmpty(),
            modifier = Modifier
                .fillMaxWidth()
                .heightIn(min = Tokens.minimumTouchTarget),
        ) {
            Text(strings.changePassword)
        }
    }
}

@Composable
private fun TwoFactorSection(
    state: AccountState,
    strings: AccountStrings,
    onDisable: (String) -> Unit,
) {
    var code by remember { mutableStateOf("") }

    Section(strings.twoFactor) {
        Text(strings.twoFactorOn, color = klinikColor("success"))

        if (!state.canDisableTwoFactor) {
            // Said rather than shown as a disabled switch: staff cannot turn
            // it off at all, and a greyed control invites somebody to keep
            // trying.
            Text(strings.twoFactorMandatory, color = klinikColor("textSecondary"))

            return@Section
        }

        Text(strings.disableTwoFactorHint, color = klinikColor("textSecondary"))

        OutlinedTextField(
            value = code,
            onValueChange = { code = it },
            label = { Text(strings.twoFactorCode) },
            modifier = Modifier.fillMaxWidth(),
        )

        OutlinedButton(
            onClick = {
                onDisable(code)
                code = ""
            },
            enabled = !state.busy && code.isNotBlank(),
            modifier = Modifier
                .fillMaxWidth()
                .heightIn(min = Tokens.minimumTouchTarget),
        ) {
            Text(strings.disableTwoFactor)
        }
    }
}

@Composable
private fun SessionsSection(
    state: AccountState,
    strings: AccountStrings,
    onEndSession: (SessionSummary) -> Unit,
    onSignOutEverywhere: () -> Unit,
) {
    var confirming by remember { mutableStateOf(false) }

    Section(strings.security) {
        state.currentSession?.let { session ->
            Row(modifier = Modifier.fillMaxWidth()) {
                Text(
                    session.deviceName ?: strings.unknownDevice,
                    color = klinikColor("textPrimary"),
                    modifier = Modifier.weight(1f),
                )
                Text(strings.thisDevice, color = klinikColor("textSecondary"))
            }
        }

        Text(strings.otherDevices, color = klinikColor("textSecondary"))

        if (state.otherSessions.isEmpty()) {
            Text(strings.noOtherDevices, color = klinikColor("textSecondary"))
        }

        state.otherSessions.forEach { session ->
            Column(modifier = Modifier.fillMaxWidth()) {
                Text(
                    session.deviceName ?: strings.unknownDevice,
                    color = klinikColor("textPrimary"),
                )
                Text(
                    strings.lastSeen(session.lastSeenAt.take(10)),
                    fontSize = Tokens.Typography.caption.size,
                    color = klinikColor("textSecondary"),
                )
                TextButton(
                    onClick = { onEndSession(session) },
                    enabled = !state.busy,
                    modifier = Modifier.heightIn(min = Tokens.minimumTouchTarget),
                ) {
                    Text(strings.revoke)
                }
            }
        }

        if (confirming) {
            // Two steps, because it takes this device with it.
            Text(strings.signOutEverywhereConfirm, color = klinikColor("warning"))
        }

        OutlinedButton(
            onClick = { if (confirming) onSignOutEverywhere() else confirming = true },
            enabled = !state.busy,
            modifier = Modifier
                .fillMaxWidth()
                .heightIn(min = Tokens.minimumTouchTarget),
        ) {
            Text(strings.signOutEverywhere)
        }
    }
}

@Composable
private fun Section(title: String, content: @Composable ColumnScope.() -> Unit) {
    Surface(color = klinikColor("surface"), modifier = Modifier.fillMaxWidth()) {
        Column(
            modifier = Modifier.padding(Tokens.Spacing.lg),
            verticalArrangement = Arrangement.spacedBy(Tokens.Spacing.sm),
        ) {
            Text(
                title,
                fontSize = Tokens.Typography.subheading.size,
                fontWeight = Tokens.Typography.subheading.weight,
                color = klinikColor("textPrimary"),
                modifier = Modifier.semantics { heading() },
            )

            content()
        }
    }
}

@Composable
private fun Centered(content: @Composable () -> Unit) {
    Box(
        modifier = Modifier.fillMaxSize().padding(Tokens.Spacing.xl),
        contentAlignment = Alignment.Center,
    ) {
        Column(horizontalAlignment = Alignment.CenterHorizontally) { content() }
    }
}
